File: 240109-search-sort/2-15-task1.py
"""
В соревновании по поеданию пончиков принимают участие N человек. Чтобы выиграть, вам нужно съесть как можно больше пончиков за 10 минут. Отображение на экране результата 3-х победителей и худшего результата. (Подсказка: программа выделяет динамическую память для массива из n элементов и запрашивает массив с клавиатуры. Попробуйте отсортировать данные разными методами. Случай, когда несколько участников съели одинаковое количество пончиков, рассматриваться не должен).
"""



def bubble_sort(nums):
    swapped = True
    while swapped:
        swapped = False
        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                nums[i], nums[i + 1] = nums[i + 1], nums[i]
                swapped = True
 
 
def selection_sort(nums):
    for i in range(len(nums)):
        lowest_value_index = i
        for j in range(i + 1, len(nums)):
            if nums[j] < nums[lowest_value_index]:
                lowest_value_index = j
        nums[i], nums[lowest_value_index] = nums[lowest_value_index], nums[i]
 
 
def insertion_sort(nums):
    for i in range(1, len(nums)):
        item_to_insert = nums[i]
        j = i - 1
        while j >= 0 and nums[j] > item_to_insert:
            nums[j + 1] = nums[j]
            j -= 1
        nums[j + 1] = item_to_insert
 
 
bidders_num = int(input("Enter the number of bidders: "))
results = []
for n in range(bidders_num):
    results.append(int(input("Enter the number of donuts eaten by participant {}: ".format(n+1))))
 
sort_types = {
    1: bubble_sort,
    2: selection_sort,
    3: insertion_sort,
}
 
results = list(set(results))
 
sort_type = int(input("Enter sort type:\n"
                      "1 - Outputting the result using Bubble sort\n"
                      "2 - Outputting the result using Sort by selection method\n"
                      "3 - Outputting the result using Insert sort\n"))
 
sort_types[sort_type](results)
if len(results) > 2:
    print("First place went to the participant who ate {} donuts".format(results[-1]))
    print("Second place went to the participant who ate {} donuts".format(results[-2]))
    print("Third place went to the participant who ate {} donuts".format(results[-3]))
    print("The worst result corresponds to {} donuts".format(results[0]))
else:
    print("You've entered less than 3 unique results")

File: 240109-search-sort/2-15-task1.c
/*
В соревновании по поеданию пончиков принимают участие N человек. Чтобы выиграть, вам нужно съесть как можно больше пончиков за 10 минут. Отображение на экране результата 3-х победителей и худшего результата. (Подсказка: программа выделяет динамическую память для массива из n элементов и запрашивает массив с клавиатуры. Попробуйте отсортировать данные разными методами. Случай, когда несколько участников съели одинаковое количество пончиков, рассматриваться не должен).
*/



#include <stdio.h>
 
int Menu();
void BubbleSort(int[], int);
void selectionSort(int[], int);
void InsertSort(int[], int);
void printArray(int[], int);
 
int main() 
{
    int dounts[20];
    int size;
 
    printf("Enter the number of bidders:(more than 3): ");
    scanf("%d", &size);
    printf("\n");
    while (size < 3) {
        printf("The number of participants must be more than 3!\n");
        printf("Enter again.\n");
        scanf("%d", &size);
    }
 
    for (int i = 0; i < size; i++) 
    {
        printf("Enter the number of donuts eaten by %d participant: ", i + 1);
        scanf("%d", &dounts[i]);
    }
 
    switch (Menu())
    {
        case 1: 
            BubbleSort(dounts, size); break;
        case 2:
            InsertSort(dounts, size); break;
        case 3: 
            selectionSort(dounts, size); break;
        default:
            printf("Incorrect choice.\n");
    }
 
    printArray(dounts, size);
    return 0;
}
 
void InsertSort(int array[], int size) 
{
    int temp, j;
    for (int i = 1; i < size; i++)
    {
        temp = array[i];
        j = i - 1;
        while (array[j] < temp && j >= 0)
        {
            array[j + 1] = array[j];
            j -= 1;
        }
        array[j + 1] = temp;
    }
}
 
void BubbleSort(int array[], int size) 
{
    int temp;
    for (int i = 0; i < size-1; i++) {
        for (int j = i; j < size; j++) {
            if (array[i] < array[j]) {
                temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
    }
}
 
void selectionSort(int array[], int size){
    int min, temp;
    for (int i = 0; i < size - 1; i++) {
        min = i;
        for (int j = i + 1; j < size; j++) {
            if (array[j] > array[min]) {
                min = j;
            }
        }
        temp = array[i];
        array[i] = array[min];
        array[min] = temp;
    }
}
 
void printArray(int array[], int size) {
    int flag = 0;
    int temp, i, how;
    for (i = 0; i < size && flag < 3; i++) {
        if (array[i] == array[i + 1]) {
            how = 1;
            temp = array[i];
            while (temp == array[i + 1]) {
                how++;
                ++i;
            }
            printf("%d participants took %d place with result - %d donuts.", how, ++flag, array[i]);
            printf("\n");
        }
        else {
            printf("The result of the participant who took %d place - %d donuts.\n", ++flag, array[i]);
        }
    }
    if ((i != size) && (size > 3)) {
        printf("The worst result corresponds to %d donuts.\n", array[size - 1]);
    }
    printf("\n");
}
 
int Menu() {
    int choice = 0;
    printf("\t1 - Outputting the result using Bubble sort.\n");
    printf("\t2 - Outputting the result using Insert sort.\n");
    printf("\t3 - Outputting the result using Selection sort.\n");
    scanf("%d", &choice);
    while (choice<1 || choice >3){
        printf("Incorrect choice. Enter again:");
        scanf("%d", &choice);
    }
    return choice;
}

